selfservice/strategy/oidc/provider_microsoft.go

```diff
@@ -0,0 +1,98 @@
+package oidc
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/dgrijalva/jwt-go"
+
+	gooidc "github.com/coreos/go-oidc"
+	"github.com/ory/herodot"
+	"github.com/ory/x/stringslice"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+type ProviderMicrosoft struct {
+	*ProviderGenericOIDC
+}
+
+func NewProviderMicrosoft(
+	config *Configuration,
+	public *url.URL,
+) *ProviderMicrosoft {
+	return &ProviderMicrosoft{
+		ProviderGenericOIDC: &ProviderGenericOIDC{
+			config: config,
+			public: public,
+		},
+	}
+}
+
+func (m *ProviderMicrosoft) OAuth2(ctx context.Context) (*oauth2.Config, error) {
+	if len(strings.TrimSpace(m.config.Tenant)) == 0 {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("No Tenant specified for the `microsoft` oidc provider %s", m.config.ID))
+	}
+
+	endpointPrefix := "https://login.microsoftonline.com/" + m.config.Tenant
+	endpoint := oauth2.Endpoint{
+		AuthURL:  endpointPrefix + "/oauth2/v2.0/authorize",
+		TokenURL: endpointPrefix + "/oauth2/v2.0/token",
+	}
+
+	scope := m.config.Scope
+	if !stringslice.Has(scope, gooidc.ScopeOpenID) {
+		scope = append(scope, gooidc.ScopeOpenID)
+	}
+
+	return &oauth2.Config{
+		ClientID:     m.config.ClientID,
+		ClientSecret: m.config.ClientSecret,
+		Endpoint:     endpoint,
+		Scopes:       scope,
+		RedirectURL:  m.config.Redir(m.public),
+	}, nil
+}
+
+func (m *ProviderMicrosoft) Claims(ctx context.Context, exchange *oauth2.Token) (*Claims, error) {
+	raw, ok := exchange.Extra("id_token").(string)
+	if !ok || len(raw) == 0 {
+		return nil, errors.WithStack(ErrIDTokenMissing)
+	}
+
+	parser := new(jwt.Parser)
+	unverifiedClaims := microsoftUnverifiedClaims{}
+	if _, _, err := parser.ParseUnverified(raw, &unverifiedClaims); err != nil {
+		return nil, err
+	}
+
+	p, err := gooidc.NewProvider(context.Background(), unverifiedClaims.Issuer)
```

Comment:
I believe this opens a very serious and real attack surface. Using the issuer from the JWT to verify the JWT means that an attacker can easily forge a JWT with his own issuer URL which has all the information (link cryptographic keys). This can be used to impersonate users.

The IssuerURL **must** come from the configuration.

Reply:
Hmm, that is true. If an attacker spoofs the `https://login.microsoftonline.com/.../token` endpoint to return a token that contains an issuer pointing to a completely unrelated OIDC provider, that would be verified as valid.

What if we verify that the issuer is on the form `https://login.microsoftonline.com/{tenant-id}/v2.0`? The `tenant-id` is in the claims, so we can use that.

That would ofcourse still potentially be an attack vector, but as far as I can see, the same vulnerability (spoofing an issuer) would be present on any OIDC provider.

Reply:
We don't really need to use the issuer url from the claims at all, the only thing that is needed for the multi-tenant flows to work is to get the actual tenant id from the JWT. So that would be another (pretty much the same) solution.

Reply:
If we don't need to use the claims issuer, then let's just use the "hardcoded" issuer! I think that's the safest. Extracting the tenant-id from the claims is totally valid, as long as the validation chain is kept in tact (which is ensured by hardcoding/hard-configuring the issuer URL)

Reply:
Sorry, forgot about this for a little while, and I my wording wasn't the best in my previous reply. What I meant when I said "we don't really need to use the issuer url from the claims", is that in stead of accepting the issuer url blindly, we can generate it using a template like `https://login.microsoftonline.com/{tenant-id}/v2.0` where `tenant-id` is present in the claims.

The tenant id will always be a guid, so by doing a bit of validation on the string from the claims. Does that sound like a viable option?

As far as I can tell, this is the only way we can get it to work with the multi-tenant flows for Microsoft. The initial auth and token endpoints are known in advance, but the issuer (with the keys to verify the JWT) is not known until the user has actually signed in with an account. So, in a way, it's not really an implementation of OIDC, but more like a mixture of OAuth2.0 + a bit of OIDC at the end.

Reply:
Ok, sounds good to me! :)